### ISSUE TITLE
No warnings for the use of uninitialized variables in several contexts

### ISSUE THREAD

**Issue by ch3root** [TrustInSoft/tis-interpreter#17]
Source code:
```c
int main()
{
  long long ll;
  (long)ll; // (1)
  
  int x;
  int y = x; // (2)
  x; // (3)
  return x; // (4)
}
```

tis-interpreter (c82b0fdfc2fb4ba9bdec9b59f935b49d6059c0bc) output:
```
[value] Analyzing a complete application starting at main
[value] Computing initial state
[value] Initial state computed
[value] done for function main
```

I think there should be a warning at least in return from main (4). Other contexts are more controversial.

For comparison:

gcc 6 output (with -Wno-unused-value -Wno-unused-variable):
```
uninit-var.c: In function ‘main’:
uninit-var.c:7:7: warning: ‘x’ is used uninitialized in this function [-Wuninitialized]
   int y = x; // (2)
       ^
```

clang 3.9 output (with -Wno-unused-value -Wno-unused-variable):
```
uninit-var.c:4:9: warning: variable 'll' is uninitialized when used here [-Wuninitialized]
  (long)ll; // (1)
        ^~
uninit-var.c:3:15: note: initialize the variable 'll' to silence this warning
  long long ll;
              ^
               = 0
uninit-var.c:7:11: warning: variable 'x' is uninitialized when used here [-Wuninitialized]
  int y = x; // (2)
          ^
uninit-var.c:6:8: note: initialize the variable 'x' to silence this warning
  int x;
       ^
        = 0
2 warnings generated.
```

**Comment by ch3root** [comment]
Another case is memcpy.

Source code:
```c
#include <string.h>

int main()
{
  int x, y;
  memcpy(&x, &y, 1);
}
```

tis-interpreter (c82b0fdfc2fb4ba9bdec9b59f935b49d6059c0bc) output:
```
[value] Analyzing a complete application starting at main
[value] Computing initial state
[value] Initial state computed
[value] done for function main
```

gcc 6 output:
```
memcpy-uninit.c: In function ‘main’:
memcpy-uninit.c:6:3: warning: ‘y’ is used uninitialized in this function [-Wuninitialized]
   memcpy(&x, &y, 1);
   ^~~~~~~~~~~~~~~~~
```

clang 3.9 output:
```
```

**Comment by pascal-cuoq** [comment]
Interestingly, not warning for (2) is by design, although it's the one Clang and GCC both warn for: there are too many blitting functions that copy (possibly uninitialized, like padding or unused members in a discriminated union) memory word by word; these functions would be detected as doing something assuredly illegal if (2) was not allowed.

I discussed this approach with Xavier Leroy a long time ago and I understand that CompCert now formalizes uninitialized memory in the same way, allowing copies of uninitialized memory by lvalue-to-lvalue assignments without an intervening conversion, and otherwise forbidding uses of uninitialized memory.

(1) and (3) should cause an error.
(4) is a little bit like an lvalue-to-lvalue assignment (especially since `x` has type `int` and there is no conversion), so it is possible to decide either way.

**Comment by pascal-cuoq** [comment]
TL;DR: if you want tis-interpreter to warn that a variable `x` is used uninitialized, use it in a computation, such as `x + 0`.

The way we handle the examples in this report has a long history. The point to which we have arrived should be better documented somewhere.

For a while, `e;` where `e` is an expression without side-effects was just ignored. In fact, there isn't even a convenient construct to remember such a statement in the AST.

We noticed that this was a problem if `e` contained UB, so the next step was to keep it in the AST. Rather than adding a new type of node, which would have broken every Frama-C analysis, we started to insert `e` in the AST as if it had been written `if (e);`, which worked for integer types, pointer types and floating-point types, allowed an analyzer to warn about UB in `e`, and did not have any effect on the program. We only did this if `e` was “dangerous”, that is if it contained a memory access.

Then we noticed that there were many more ways to be dangerous than to contain memory accesses. We widened the definition of “dangerous” to include accesses to simple variables, since such a variable could be uninitialized. At that time the analyzer began to warn about your example (3).

Then we noticed that an lvalue can be a struct, and that `if (s)` where `s` is a struct is an invalid construct. So we changed the translation of `e` to `type_of_e tmp = e;`.

This made the warning for (3) disappear, since we do not warn for lvalue-to-lvalue assignments that do not incur a conversion. But that works out pretty well in practice.

Being silent on (1) is a bug, because there is a conversion.
